portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -0,0 +1,79 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.PrintWriter;
+import com.google.gson.Gson;
+import com.google.sps.data.Comment;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet responsible for listing tasks. */
+@WebServlet("/login_page")
+public class LoginServlet extends HttpServlet {
+
+  private String loginStatus = "false";
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json;");
+    PrintWriter out = response.getWriter();
+
+    // Only logged-in users can see the form
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      loginStatus = "true";
+      String userEmail = userService.getCurrentUser().getEmail();
+      String urlToRedirectToAfterUserLogsOut = "/index.html";
+      String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+      String message_1 = "<p>Hello " + userEmail + "!</p>" 
+           + "<p>Logout <a href=\"" + logoutUrl + "\">here</a>.</p>";
+      List<String> pair_1 = new ArrayList<String>();
```

Comment:
Passing these values in a list works. But I think there's a better pattern available to you. The problem with lists is that you then have to parse these values on the client and rely on the order of the elements. This is a bit obscure. Instead, you could write a Java object with named fields (`loginStatus` and `message`) and serialize this object into JSON. Then you can refer to elements in the object _by name_. Also loginStatus should be a boolean, so using a list of strings doesn't work with a boolean and a String. Using an object has no such limitations.

Reply:
Thanks! I tried to use a Pair but for some reason couldn't import it

Reply:
You can create a dedicated Java class https://github.com/google/gson/blob/master/UserGuide.md#object-examples